### ISSUE TITLE
Nested generic method calls can hide null value

### ISSUE THREAD

**Issue by twistedsquare** [typetools/checker-framework#2235]
I have some code which passes the nullness checker but will cause a NullPointerException when run.  I've boiled it down to the following minimal example.  The key seems to be the nested calls to the generic `make` method.  I think perhaps the nullness checker is inferring a `@Nullable` type for the inner call, but then it should surely propagate this and not allow the assignment to a non-null value.  Here's the code:

```
public class TestNest
{
    // Simple wrapper class with a public generic method to make an instance:
    static class Holder<T>
    {
        T t;

        private Holder(T t)
        {
            this.t = t;
        }
        
        public static <T> Holder<T> make(T t)
        {
            return new Holder<>(t);
        }
    }
    
    public static void main(String[] args) throws Exception
    {
        // Null is hidden via nested calls, but assigned to a non-null type:
        Holder<Holder<String>> h = Holder.make(Holder.make(null));
        // NullPointerException will fire here:
        h.t.t.toString();
    }
}
```

Tested on 2.5.8.

**Comment by wmdietl** [comment]
This false negative is unfortunately caused by #979. At the moment for such nested method type argument inference we infer incorrect type arguments and suppress errors to reduce the number of false positives.

We do support the `-AconservativeUninferredTypeArguments` flag, which gives you errors for this invocation:

```
./checker/bin-devel/javac -processor nullness Issue2235.java -AconservativeUninferredTypeArguments
Issue2235.java:21: error: [assignment.type.incompatible] incompatible types in assignment.
        Holder<Holder<String>> h = Holder.make(Holder.make(null));
                                              ^
  found   : @Initialized @NonNull Holder<@Initialized @NonNull Holder<? extends @Initialized @Nullable Object>>
  required: @UnknownInitialization @Nullable Holder<@Initialized @NonNull Holder<@Initialized @NonNull String>>
Issue2235.java:21: error: [type.argument.type.incompatible] incompatible types in type argument.
        Holder<Holder<String>> h = Holder.make(Holder.make(null));
                                                          ^
  found   : ? extends Object
  required: [extends Object super null]
Issue2235.java:21: error: [argument.type.incompatible] incompatible types in argument.
        Holder<Holder<String>> h = Holder.make(Holder.make(null));
                                                           ^
  found   : @UnknownKeyFor null
  required: ?[ extends @UnknownKeyFor Object super @UnknownKeyFor Void]
3 errors
```

However, it might also give you false positives and might get annoying.
Can you provide feedback in how well the conservative option works?

**Comment by twistedsquare** [comment]
I took the plunge and turned on the conservative checker.  It ended up not spotting any further null issues (after I fixed the one that triggered this bug) but I feel better for having it on.  The main places where I needed to add a line was all the calls to Collectors.toList plus various nested generic building calls (e.g. Stream.of).  In case it helps, I totalled up, and to compile under the conservative checker, I had to modify 0.5% of the lines in my project which uses more generics than average.  Which was a bit of a hit to do all at once, but it's a proportion I can live with.  IntelliJ's Alt-Enter menu item "add explicit type arguments" (when it was available) sped things up a lot.

**Comment by mernst** [comment]
Thanks very much for the bug report.  We really appreciate it.

Also, thanks for reporting your experience with `-AconservativeUninferredTypeArguments`.  I had expected use of it to be more disruptive, and it's great to hear that my pessimism was unfounded.
